File: app/utils/dawproject_python/channel.py
from lxml import etree as ET
from classes.lane import Lane
from classes.realParameter import RealParameter
from classes.boolParameter import BoolParameter
from classes.device import Device


class Channel(Lane):
    def __init__(
        self,
        role=None,
        audio_channels=2,
        volume=None,
        pan=None,
        mute=None,
        solo=None,
        destination=None,
        sends=None,
        devices=None,
        name=None,
        color=None,
        comment=None,
    ):
        super().__init__(name, color, comment)
        self.role = role
        self.audio_channels = audio_channels
        self.volume = volume
        self.pan = pan
        self.mute = mute
        self.solo = solo
        self.destination = destination
        self.sends = sends if sends else []
        self.devices = devices if devices else []

    def to_xml(self):
        # Inherit XML generation from Lane and modify tag
        channel_elem = super().to_xml()
        channel_elem.tag = "Channel"

        # Set attributes for the Channel element
        if self.role is not None:
            channel_elem.set("role", str(self.role))  # Ensure role is a string
        if self.audio_channels is not None:
            channel_elem.set(
                "audioChannels", str(self.audio_channels)
            )  # Convert int to string
        if self.solo is not None:
            channel_elem.set(
                "solo", str(self.solo).lower()
            )  # Convert boolean to lowercase string
        if self.destination:
            channel_elem.set(
                "destination", str(self.destination.id)
            )  # Assuming destination has an id attribute

        # Append complex elements if they exist
        if self.devices:
            devices_elem = ET.SubElement(channel_elem, "Devices")
            for device in self.devices:
                devices_elem.append(device.to_xml())
        if self.mute:
            mute_elem = ET.SubElement(channel_elem, "Mute")
            mute_elem.append(self.mute.to_xml())
        if self.pan:
            pan_elem = ET.SubElement(channel_elem, "Pan")
            pan_elem.set(
                "value", str(self.pan.value) if self.pan.value is not None else ""
            )
            pan_elem.set(
                "unit", str(self.pan.unit.value) if self.pan.unit is not None else ""
            )
        if self.sends:
            sends_elem = ET.SubElement(channel_elem, "Sends")
            for send in self.sends:
                send_elem = ET.SubElement(sends_elem, "Send")
                send_elem.append(send.to_xml())
        if self.volume:
            volume_elem = ET.SubElement(channel_elem, "Volume")
            volume_elem.set(
                "value", str(self.volume.value) if self.volume.value is not None else ""
            )
            volume_elem.set(
                "unit",
                str(self.volume.unit.value) if self.volume.unit is not None else "",
            )

        return channel_elem

    @classmethod
    def from_xml(cls, element):
        from classes.send import Send

        instance = super().from_xml(element)

        instance.role = element.get("role")

        audio_channels = element.get("audioChannels")
        instance.audio_channels = (
            int(audio_channels) if audio_channels is not None else 2
        )

        volume_elem = element.find("Volume")
        instance.volume = (
            RealParameter.from_xml(volume_elem) if volume_elem is not None else None
        )

        pan_elem = element.find("Pan")
        instance.pan = (
            RealParameter.from_xml(pan_elem) if pan_elem is not None else None
        )

        mute_elem = element.find("Mute")
        instance.mute = (
            BoolParameter.from_xml(mute_elem) if mute_elem is not None else None
        )

        solo = element.get("solo")
        instance.solo = solo.lower() == "true" if solo else None

        destination_id = element.get("destination")
        instance.destination = (
            Channel.get_channel_by_id(destination_id)
            if destination_id is not None
            else None
        )

        sends = []
        for send_elem in element.findall(".//Sends/Send"):
            sends.append(Send.from_xml(send_elem))
        instance.sends = sends

        devices = []
        for device_elem in element.findall(".//Devices/*"):
            devices.append(Device.from_xml(device_elem))
        instance.devices = devices

        return instance

    @staticmethod
    def get_channel_by_id(channel_id):
        # Implement logic to retrieve a Channel instance by its ID
        pass

File: app/utils/dawproject/channel.ts
import { BoolParameter } from "./boolParameter";
import { Device } from "./device/device";
import { Lane } from "./lane";
import { MixerRole } from "./mixerRole";
import { RealParameter } from "./realParameter";
import { Referenceable } from "./referenceable";
import { DeviceRegistry } from "./registry/deviceRegistry";
import { registerLane } from "./registry/laneRegistry";
import { Send } from "./send";
import { IChannel, IDevice } from "./types";
import { Utility } from "./utility";

/**
 * Represents a mixer channel. It provides the ability to route signals to other channels and can contain
 * Device/Plug-in for processing.
 */

const channelFactory = (xmlObject: any): Channel => {
  const instance = new Channel();
  instance.fromXmlObject(xmlObject);
  return instance;
};

@registerLane("Channel", channelFactory)
export class Channel extends Lane implements IChannel {
  /** Role of this channel in the mixer. */
  role?: MixerRole;
  /** Number of audio-channels of this mixer channel. (1=mono, 2=stereo…) */
  audioChannels?: number;
  /** Channel volume */
  volume?: RealParameter;
  /** Channel pan/balance */
  pan?: RealParameter;
  /** Channel mute */
  mute?: BoolParameter;
  /** Channel solo */
  solo?: boolean;
  /** Output channel routing */
  destination?: IChannel; // Reference to another Channel
  /** Send levels & destination */
  sends: Send[];
  /** Devices & plug-ins of this channel */
  devices: IDevice[];

  constructor(
    role?: MixerRole,
    audioChannels: number = 2,
    volume?: RealParameter,
    pan?: RealParameter,
    mute?: BoolParameter,
    solo?: boolean,
    destination?: Channel,
    sends?: Send[],
    devices?: Device[],
    name?: string,
    color?: string,
    comment?: string
  ) {
    super(name, color, comment);
    this.role = role;
    this.audioChannels = audioChannels;
    this.volume = volume;
    this.pan = pan;
    this.mute = mute;
    this.solo = solo;
    this.destination = destination;
    this.sends = sends || [];
    this.devices = devices || [];
  }

  toXmlObject(): any {
    const obj: any = {
      Channel: {
        ...super.toXmlObject(),
      },
    };

    // add optional attributes
    Utility.addAttribute(obj.Channel, "role", this);
    Utility.addAttribute(obj.Channel, "audioChannels", this);
    Utility.addAttribute(obj.Channel, "solo", this);
    Utility.addAttribute(obj.Channel, "destination", this, {
      sourceProperty: "destination.id",
    });

    // Add children directly
    const groupedDevices = Utility.groupChildrenByTagName(this.devices);
    if (groupedDevices) {
      obj.Channel.Devices = groupedDevices;
    }

    // Add optional elements, each in their proper XML element form
    if (this.mute) {
      obj.Channel.Mute = this.mute.toXmlObject().BoolParameter;
    }
    if (this.pan) {
      obj.Channel.Pan = this.pan.toXmlObject().RealParameter;
    }
    if (this.volume) {
      obj.Channel.Volume = this.volume.toXmlObject().RealParameter;
    }
    if (this.sends && this.sends.length > 0) {
      obj.Channel.Sends = {
        Send: this.sends.map(send => send.toXmlObject().Send),
      };
    }

    return obj;
  }

  fromXmlObject(xmlObject: any): this {
    super.fromXmlObject(xmlObject); // populate inherited attributes from Lane

    // handle optional attributes
    Utility.populateAttribute<MixerRole>(xmlObject, "role", this, {
      castTo: MixerRole,
    });

    Utility.populateAttribute<number>(xmlObject, "audioChannels", this, {
      castTo: Number,
      validator: v => v > 0,
    });

    Utility.populateAttribute<boolean>(xmlObject, "solo", this, {
      castTo: Boolean,
    });

    // Handle optional elements
    if (xmlObject.Volume) {
      this.volume = new RealParameter().fromXmlObject(xmlObject.Volume);
    }

    if (xmlObject.Pan) {
      this.pan = new RealParameter().fromXmlObject(xmlObject.Pan);
    }

    if (xmlObject.Mute) {
      this.mute = new BoolParameter().fromXmlObject(xmlObject.Mute);
    }

    // read destination ID and look it up in the Referenceable registry
    const destinationId = xmlObject["@_destination"];
    if (destinationId) {
      const destination = Referenceable.getById(destinationId);
      // Check if the retrieved object is a Channel before assigning
      if (destination instanceof Channel) {
        this.destination = destination;
      } else {
        console.warn(
          `Retrieved object with ID ${destinationId} is not a Channel and cannot be assigned as a destination.`
        );
        this.destination = undefined; // Ensure destination is undefined if not a Channel
      }
    }

    // Handle sends array with proper error handling
    this.sends = [];
    if (xmlObject.Sends?.Send) {
      const sendArray = Array.isArray(xmlObject.Sends.Send)
        ? xmlObject.Sends.Send
        : [xmlObject.Sends.Send];

      this.sends = sendArray.map((sendObj: any) => {
        try {
          return new Send().fromXmlObject(sendObj);
        } catch (error: unknown) {
          const message =
            error instanceof Error ? error.message : String(error);
          throw new Error(`Failed to parse Send: ${message}`);
        }
      });
    }

    const devices: IDevice[] = [];
    if (xmlObject.Devices) {
      // Iterate through all properties in xmlObject.Devices to find Device elements
      for (const tagName in xmlObject.Devices) {
        // Skip attributes (those starting with @_)
        if (tagName.startsWith("@_")) continue;

        const DeviceClass = DeviceRegistry.getDeviceClass(tagName);
        if (DeviceClass) {
          const deviceData = xmlObject.Devices[tagName];
          const deviceArray = Array.isArray(deviceData)
            ? deviceData
            : [deviceData];

          for (const deviceObj of deviceArray) {
            try {
              // Create a new instance and call fromXmlObject on it
              const deviceInstance = new DeviceClass().fromXmlObject(deviceObj);
              devices.push(deviceInstance);
            } catch (e) {
              console.error(
                `Error deserializing nested device element ${tagName} in Channel:`,
                e
              );
            }
          }
        } else {
          console.warn(
            `Skipping deserialization of unknown nested device element in Channel: ${tagName}`
          );
        }
      }
    }
    this.devices = devices;

    return this;
  }
}
